File: joiner-core/src/main/java/cz/encircled/joiner/core/Joiner.java
package cz.encircled.joiner.core;

import com.querydsl.core.types.*;
import cz.encircled.joiner.core.vendor.EclipselinkRepository;
import cz.encircled.joiner.core.vendor.HibernateRepository;
import cz.encircled.joiner.core.vendor.JoinerJpaQuery;
import cz.encircled.joiner.core.vendor.VendorRepository;
import cz.encircled.joiner.exception.AliasMissingException;
import cz.encircled.joiner.exception.JoinerException;
import cz.encircled.joiner.exception.JoinerExceptions;
import cz.encircled.joiner.query.JoinerQuery;
import cz.encircled.joiner.query.QueryFeature;
import cz.encircled.joiner.query.QueryOrder;
import cz.encircled.joiner.query.join.J;
import cz.encircled.joiner.query.join.JoinDescription;
import cz.encircled.joiner.query.join.JoinGraphRegistry;
import cz.encircled.joiner.util.JoinerUtils;
import jakarta.persistence.EntityManager;
import jakarta.persistence.Query;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.AnnotatedElement;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static cz.encircled.joiner.util.Assert.notNull;

/**
 * Base class of Joiner. Contains basic database operations.
 * <p>
 * In spring-based environment can be instantiated using spring JoinerConfiguration.
 * </p>
 * <p>
 * For repository-per-entity approach this class should not be accessed directly. Instead, repositories may implement {@link JoinerRepository}.
 * </p>
 *
 * @author Kisel on 26.01.2016.
 */
public class Joiner {

    private static final Logger log = LoggerFactory.getLogger(Joiner.class);

    private final EntityManager entityManager;

    private final VendorRepository vendorRepository;

    private JoinGraphRegistry joinGraphRegistry;

    private final AliasResolver aliasResolver;

    private final PredicateAliasResolver predicateAliasResolver = new DefaultPredicateAliasResolver(this);

    private JoinerProperties joinerProperties;

    public Joiner(EntityManager entityManager) {
        this(entityManager, new JoinerProperties());
    }

    public Joiner(EntityManager entityManager, JoinerProperties joinerProperties) {
        notNull(entityManager);

        this.joinerProperties = joinerProperties;
        this.entityManager = entityManager;
        this.aliasResolver = new DefaultAliasResolver(entityManager);

        String implName = entityManager.getDelegate().getClass().getName();
        this.vendorRepository = implName.startsWith("org.hibernate") ? new HibernateRepository() : new EclipselinkRepository();
    }

    public Joiner withProperties(JoinerProperties props) {
        return new Joiner(entityManager, props);
    }

    /**
     * Execute a query and return a result or null.
     */
    public <T, R> R findOne(JoinerQuery<T, R> request) {
        List<R> list = find(request);
        if (list.isEmpty()) {
            return null;
        } else if (list.size() == 1) {
            return list.get(0);
        } else {
            throw JoinerExceptions.multipleEntitiesFound();
        }
    }

    /**
     * Execute a query and return the List of results.
     */
    public <T, R> List<R> find(JoinerQuery<T, R> request) {
        preprocessRequestQuery(request);

        List<R> result;
        try (JoinerJpaQuery jpaQuery = vendorRepository.createQuery(request, joinerProperties, entityManager)) {
            Query finalQuery = jpaQuery.jpaQuery;
            for (QueryFeature feature : getQueryFeatures(request)) {
                finalQuery = feature.after(request, finalQuery);
            }
            result = vendorRepository.fetchResult(request, finalQuery);
        }

        for (QueryFeature queryFeature : getQueryFeatures(request)) {
            queryFeature.postLoad(request, result);
        }

        return result;
    }

    /**
     * <p>
     * Execute a query and stream results using JPA streaming.
     * Fetch size is configured using a hint <i>org.hibernate.fetchSize</i> (HibernateHints.HINT_FETCH_SIZE)
     * </p>
     * <p>
     * Using streaming disables caching and auto-flush for the query.
     * </p>
     * JPA streaming is supported only by Hibernate ORM.
     */
    public <T, R> Stream<R> findStream(JoinerQuery<T, R> request) {
        preprocessRequestQuery(request);

        Stream<R> result;
        try (JoinerJpaQuery jpaQuery = vendorRepository.createQuery(request, joinerProperties, entityManager)) {
            Query finalQuery = jpaQuery.jpaQuery;
            for (QueryFeature feature : getQueryFeatures(request)) {
                finalQuery = feature.after(request, finalQuery);
            }
            result = vendorRepository.streamResult(request, finalQuery);
        }

        return result.peek(r -> {
            for (QueryFeature queryFeature : getQueryFeatures(request)) {
                queryFeature.postLoad(request, List.of(r));
            }
        });
    }

    public <I, T extends Collection<I>> T save(T entities) {
        for (Object entity : entities) {
            save(entity);
        }
        return entities;
    }

    public <T> T save(T entity) {
        entityManager.persist(entity);
        return entity;
    }

    public <T, R> JoinerJpaQuery toJPAQuery(JoinerQuery<T, R> request) {
        preprocessRequestQuery(request);
        return vendorRepository.createQuery(request, joinerProperties, entityManager);
    }

    public <T, R> void preprocessRequestQuery(JoinerQuery<T, R> request) {
        notNull(request);
        notNull(request.getFrom());

        setJoinsFromJoinsGraphs(request);

        for (QueryFeature feature : getQueryFeatures(request)) {
            request = feature.before(request);
        }

        Set<Path<?>> usedAliases = new HashSet<>();
        usedAliases.add(request.getFrom());

        List<JoinDescription> joins = preprocessJoins(request, usedAliases);

        addJoins(joins, request);
        applyPredicates(request, usedAliases, joins);
        applyPaging(request, usedAliases, joins);
    }

    private <T, R> List<QueryFeature> getQueryFeatures(JoinerQuery<T, R> request) {
        if (!joinerProperties.defaultFeatures.isEmpty()) {
            ArrayList<QueryFeature> res = new ArrayList<>(joinerProperties.defaultFeatures);
            res.addAll(request.getFeatures());
            return res;
        }
        return request.getFeatures();
    }

    private <T, R> List<JoinDescription> preprocessJoins(JoinerQuery<T, R> request, Set<Path<?>> usedAliases) {
        List<JoinDescription> joins = J.unrollChildrenJoins(request.getJoins());
        for (JoinDescription join : joins) {
            if (join.getCollectionPath() == null && join.getSingularPath() == null) {
                aliasResolver.resolveFieldPathForJoinAlias(join, request.getFrom());
            }
            usedAliases.add(join.getAlias());
        }
        for (JoinDescription join : joins) {
            if (join.getOn() != null) {
                join.on(predicateAliasResolver.resolvePredicate(join.getOn(), joins, usedAliases));
            }
        }
        return joins;
    }

    /**
     * Apply "where", "groupBy" and "having"
     */
    private <T, R> void applyPredicates(JoinerQuery<T, R> request, Set<Path<?>> usedAliases, List<JoinDescription> joins) {
        if (request.getWhere() != null) {
            Predicate where = predicateAliasResolver.resolvePredicate(request.getWhere(), joins, usedAliases);
            checkAliasesArePresent(where, usedAliases);
            request.where(where);
        }

        if (request.getGroupBy() != null) {
            Map<AnnotatedElement, List<JoinDescription>> grouped = joins.stream()
                    .collect(Collectors.groupingBy(j -> j.getOriginalAlias().getAnnotatedElement()));
            if (!request.getGroupBy().isEmpty()) {
                List<Expression<?>> mapped = request.getGroupBy().stream().map(g -> {
                    if (g instanceof Path<?> path) {
                        Path<?> grouping = predicateAliasResolver.resolvePath(path, grouped, usedAliases);
                        checkAliasesArePresent(grouping, usedAliases);
                        return grouping;
                    } else {
                        return g;
                    }
                }).toList();

                request.groupBy(mapped);
            }
        }

        if (request.getHaving() != null) {
            Predicate having = predicateAliasResolver.resolvePredicate(request.getHaving(), joins, usedAliases);
            checkAliasesArePresent(having, usedAliases);
            request.having(having);
        }
    }

    private <T, R> void applyPaging(JoinerQuery<T, R> request, Set<Path<?>> usedAliases, List<JoinDescription> joins) {
        Map<AnnotatedElement, List<JoinDescription>> grouped = joins.stream()
                .collect(Collectors.groupingBy(j -> j.getOriginalAlias().getAnnotatedElement()));

        for (QueryOrder<?> queryOrder : request.getOrder()) {
            if (queryOrder.getTarget() instanceof Path<?>) {
                Path path = predicateAliasResolver.resolvePath((Path<?>) queryOrder.getTarget(), grouped, usedAliases);
                queryOrder.setTarget(path);
            }
            checkAliasesArePresent(queryOrder.getTarget(), usedAliases);
        }
    }

    private <T, R> void setJoinsFromJoinsGraphs(JoinerQuery<T, R> request) {
        if (!request.getJoinGraphs().isEmpty()) {
            if (joinGraphRegistry == null) {
                throw new JoinerException("Join graph are set, but joinGraphRegistry is null!");
            }

            Class<? extends T> queryRootClass = request.getFrom().getType();

            for (Object name : request.getJoinGraphs()) {
                List<JoinDescription> joins = joinGraphRegistry.getJoinGraph(queryRootClass, name);

                // Fetch is not allowed in count queries
                if (request.isCount()) {
                    J.unrollChildrenJoins(joins).forEach(j -> j.fetch(false));
                }

                request.joins(joins);
            }
        }
    }

    private void addJoins(List<JoinDescription> joins, JoinerQuery<?, ?> request) {
        EntityPath<?> rootPath = request.getFrom();
        boolean doFetch = request.getFrom().equals(request.getReturnProjection());
        for (JoinDescription join : joins) {
            if (doFetch && join.isFetch()) {
                vendorRepository.addFetch(join, joins, rootPath, request);
            }
        }
    }

    private void checkAliasesArePresent(Expression<?> expression, Set<Path<?>> usedAliases) {
        for (Path<?> path : JoinerUtils.collectPredicatePaths(expression)) {
            Path<?> predicatePath = path.getRoot();
            if (!predicatePath.toString().startsWith("any(")) {
                if (!usedAliases.contains(predicatePath)) {
                    throw new AliasMissingException("Alias " + predicatePath + " is not present in joins!");
                }
            }
        }
    }

    public void setJoinGraphRegistry(JoinGraphRegistry joinGraphRegistry) {
        this.joinGraphRegistry = joinGraphRegistry;
    }

    public JoinerProperties getJoinerProperties() {
        return joinerProperties;
    }

    public void setJoinerProperties(JoinerProperties joinerProperties) {
        this.joinerProperties = Objects.requireNonNullElse(joinerProperties, new JoinerProperties());
    }

    AliasResolver getAliasResolver() {
        return aliasResolver;
    }

}

File: joiner-core/src/main/java/cz/encircled/joiner/core/Joiner.kt
package cz.encircled.joiner.core

import com.google.common.collect.ArrayListMultimap
import com.mysema.query.JoinType
import com.mysema.query.jpa.impl.AbstractJPAQuery
import com.mysema.query.jpa.impl.JPAQuery
import com.mysema.query.types.EntityPath
import com.mysema.query.types.Expression
import com.mysema.query.types.Operation
import com.mysema.query.types.Path
import cz.encircled.joiner.core.vendor.EclipselinkRepository
import cz.encircled.joiner.core.vendor.HibernateRepository
import cz.encircled.joiner.core.vendor.JoinerVendorRepository
import cz.encircled.joiner.exception.AliasMissingException
import cz.encircled.joiner.exception.JoinerException
import cz.encircled.joiner.query.Q
import cz.encircled.joiner.query.QueryFeature
import cz.encircled.joiner.query.join.JoinDescription
import cz.encircled.joiner.query.join.JoinGraphRegistry
import cz.encircled.joiner.util.Assert
import cz.encircled.joiner.util.ReflectionUtils
import java.util.*
import javax.persistence.EntityManager

/**
 * @author Kisel on 26.01.2016.
 */
class Joiner(private val entityManager: EntityManager) {

    private var joinerVendorRepository: JoinerVendorRepository? = null

    private var joinGraphRegistry: JoinGraphRegistry? = null

    private val aliasResolver = DefaultAliasResolver()

    init {
        val implName = entityManager.delegate.javaClass.name
        if (implName.startsWith("org.hibernate")) {
            this.joinerVendorRepository = HibernateRepository()
        } else if (implName.startsWith("org.eclipse")) {
            this.joinerVendorRepository = EclipselinkRepository()
        }
    }

    fun <T> findOne(request: Q<T>): T? {
        return findOne(request, request.from)
    }

    fun <T, P> findOne(request: Q<T>, projection: Expression<P>): P? {
        val list = find(request, projection)
        if (list.isEmpty()) {
            return null
        } else if (list.size == 1) {
            return list[0]
        } else {
            throw JoinerException("FindOne returned multiple records!")
        }
    }

    fun <T> find(request: Q<T>): List<T> {
        Assert.notNull(request)
        return find(request, request.from)
    }

    fun <T, P> find(request: Q<T>, projection: Expression<P>): List<P> {
        var request = request
        Assert.notNull(request)
        Assert.notNull(projection)
        // TODO extract validation
        Assert.notNull(request.from)

        setJoinsFromJoinsGraphs(request)

        for (feature in request.features) {
            request = doPreProcess(request, feature)
        }

        var query = joinerVendorRepository!!.createQuery(entityManager)
        makeInsertionOrderHints(query)

        query.from(request.from)
        if (request.isDistinct) {
            query.distinct()
        }

        val usedAliases = HashSet<Path<*>>()
        usedAliases.add(request.from)

        val joins = unrollChildren(request.joins)
        for (join in joins) {
            aliasResolver.resolveJoinAlias(join, request.from)
            usedAliases.add(join.alias)
        }
        addJoins(joins, query, request.from, request.from == projection)

        addHints(request, query)

        checkAliasesArePresent(request.where, usedAliases)
        checkAliasesArePresent(request.having, usedAliases)
        checkAliasesArePresent(request.groupBy, usedAliases)

        query.where(request.where)

        if (request.groupBy != null) {
            query.groupBy(request.groupBy)
        }

        if (request.having != null) {
            query.having(request.having)
        }

        for (feature in request.features) {
            query = doPostProcess(request, query, feature)
        }

        return query.list(projection)
    }

    private fun <T> setJoinsFromJoinsGraphs(request: Q<T>) {
        if (!request.joinGraphs.isEmpty()) {
            if (joinGraphRegistry == null) {
                throw JoinerException("Join graph are set, but joinGraphRegistry is null!")
            }

            val queryRootClass = request.from.type

            for (name in request.joinGraphs) {
                val joins = joinGraphRegistry!!.getJoinGraph(queryRootClass, name)
                if (joins.isEmpty()) {
                    throw JoinerException(String.format("JoinGraph with name [%s] is not defined for class [%s]", name, queryRootClass))
                } else {
                    request.joins(joins)
                }
            }
        }
    }

    private fun doPostProcess(request: Q<*>, query: JPAQuery, feature: QueryFeature): JPAQuery {
        return feature.after(request, query)
    }

    private fun <T> doPreProcess(request: Q<T>, feature: QueryFeature): Q<T> {
        return feature.before(request)
    }

    private fun makeInsertionOrderHints(sourceQuery: AbstractJPAQuery<JPAQuery>) {
        val f = ReflectionUtils.findField(AbstractJPAQuery::class.java, "hints")
        ReflectionUtils.setField(f!!, sourceQuery, ArrayListMultimap.create<Any, Any>())
    }

    private fun addJoins(joins: List<JoinDescription>, query: JPAQuery, rootPath: EntityPath<*>, doFetch: Boolean) {
        for (join in joins) {
            joinerVendorRepository!!.addJoin(query, join)
            if (doFetch && join.isFetch) {
                if (join.joinType == JoinType.RIGHTJOIN) {
                    throw JoinerException("Fetch is not supported for right join!")
                }
                joinerVendorRepository!!.addFetch(query, join, joins, rootPath)
            }
        }
    }

    private fun unrollChildren(joins: Set<JoinDescription>): List<JoinDescription> {
        val collection = LinkedList<JoinDescription>()

        for (joinDescription in joins) {
            unrollChildrenInternal(joinDescription, collection)
        }

        return collection
    }

    private fun unrollChildrenInternal(join: JoinDescription, collection: MutableList<JoinDescription>) {
        collection.add(join)
        for (child in join.children) {
            unrollChildrenInternal(child, collection)
        }
    }

    private fun addHints(request: Q<*>, query: JPAQuery) {
        for ((key, hintValues) in request.hints) {
            for (value in hintValues) {
                query.setHint(key, value)
            }
        }
    }

    private fun checkAliasesArePresent(expression: Expression<*>?, usedAliases: Set<Path<*>>) {
        for (path in collectPredicatePaths(expression)) {
            val predicatePath = path.root
            if (!predicatePath.toString().startsWith("any(")) {
                if (!usedAliases.contains(predicatePath)) {
                    throw AliasMissingException("Alias $predicatePath is not present in joins!")
                }
            }
        }
    }

    private fun collectPredicatePaths(expression: Expression<*>?): List<Path<*>> {
        val result = ArrayList<Path<*>>()
        if (expression != null) {
            collectPredicatePathsInternal(expression, result)
        }
        return result
    }

    private fun collectPredicatePathsInternal(expression: Expression<*>, paths: MutableList<Path<*>>) {
        if (expression is Path<*>) {
            paths.add(expression)
        } else if (expression is Operation<*>) {
            for (exp in expression.args) {
                collectPredicatePathsInternal(exp, paths)
            }
        }
    }

    fun setJoinGraphRegistry(joinGraphRegistry: JoinGraphRegistry) {
        this.joinGraphRegistry = joinGraphRegistry
    }

}
